speculative_decoding/packages/spec_dec.py

```diff
@@ -0,0 +1,369 @@
+        with helpers.timeit(f"Generate({self._model_name}) - input prep", skip=True):
+            inputs = helpers.make_trtllm_inputs(
+                confirmed_ids,
+                num_gen_tokens,
+                sampling_config,
+                draft_ids,
+                self._tokenizer.eos_token_id,
+                self._tokenizer.pad_token_id,
+                self._tokenize_word_list(bad_word_list),
+                self._tokenize_word_list(stop_words_list),
+            )
+        with helpers.timeit(f"Verify+Generate({self._model_name})"):
+            result = await self._client.infer(
+                self._model_name, inputs, request_id=request_id
+            )
+            output_ids = helpers.extract_trtllm_outputs(result)
+
+        with helpers.timeit(f"Generate({self._model_name}) - detokenize", skip=True):
+            output_text = self._tokenizer.decode(output_ids, skip_special_tokens=True)
+        return output_text, output_ids
+
+
+QUEUE_SENTINEL = None
+
+
+async def run_speculative_inference(
+    target_model: ModelWrapper,
+    draft_model: ModelWrapper,
+    request: helpers.GenerationRequest,
+    max_num_draft_tokens: int,
+    request_id: str,
+    result_queue: asyncio.Queue[str | QUEUE_SENTINEL] | None = None,
+    verbose: bool = False,
+    iteration_delay: float = 0.0,
+) -> SpeculationState:
+    """Runs the speculative decoding control flow loop betweeen draft and target model
+
+    * Model invocations are async inference requests to the triton server.
+    * For streaming, intermediate results are published in an optiona `result_queue`,
+       note that currently each result includes the context prefix, not just generated
+       text.
+    * For the loop to advance, it is necessary to await the results all inferernce
+      requests. But if this coroutine function is scheduled as a task, e.g.
+      by `asyncio.ensure_future`, one python server can run multiple inferences
+      concurrently. The inference requests coming from those might can be collated to
+      batches by the tritonserver if batching is configured which can increase total
+      throughput.
+    """
+    state = SpeculationState(request.prompt, target_model._tokenizer, debugging=verbose)
+    max_total_tokens = state.num_tokens + request.max_num_generated_tokens
+    num_chars_generated = 0
+    while True:
+        num_draft_tokens = min(
+            max_num_draft_tokens,
+            max_total_tokens - state.num_tokens,
+        )
+        combined_draft, _ = await draft_model.generate(
+            state.get_verified_text(),
+            num_draft_tokens,
+            request_id,
+            request.sampling_config,
+            request.bad_word_list,
+            request.stop_words_list,
+        )
+        state.update_draft(combined_draft)
+        if iteration_delay:
+            time.sleep(iteration_delay)
+
+        verification_inputs = state.get_verification_inputs()
+        if verification_inputs is None:
```

Comment:
Worth commenting when this case can be hit. It's perhaps when the draft token does not generate any new tokens in the code above in the while loop. It's not immediately clear when reading the code as to what conditions that can occur in and will be helpful to have as comment here.